File: src/reporting.ts
import type { EconomyEvent } from './events';
import type { QueryWindow } from './query';

type NullableTotal = { known: number; unknownRecords: number; total: number | null };
function sum(values: (number | null)[]): NullableTotal {
  const known = values.reduce<number>((total, value) => total + (value ?? 0), 0);
  const unknownRecords = values.filter(value => value === null).length;
  return { known, unknownRecords, total: unknownRecords > 0 || values.length === 0 ? null : known };
}

export function report(events: EconomyEvent[], window: QueryWindow, by: 'model' | 'category' | 'client' = 'model') {
  const usage = events.filter(event => event.eventType === 'usage');
  const transforms = events.filter((event): event is Extract<EconomyEvent, {eventType: 'transform'}> => event.eventType === 'transform' && event.payload.sent);
  const groups = new Map<string, typeof usage>();
  for (const event of usage) {
    const key = by === 'model' ? event.modelId ?? 'unknown' : by === 'client' ? event.clientId : event.payload.category;
    groups.set(key, [...(groups.get(key) ?? []), event]);
  }
  const reductions = new Map<string, { scope: string; evidence: string; tokenizer: string | null; values: (number | null)[] }>();
  for (const event of transforms) {
    const p = event.payload;
    const key = JSON.stringify([p.scope, event.evidence, p.tokenizer, event.modelId]);
    const group = reductions.get(key) ?? { scope: p.scope, evidence: event.evidence, tokenizer: p.tokenizer, modelId: event.modelId, values: [] as (number | null)[] };
    group.values.push(p.inputTokens === null || p.outputTokens === null ? null : p.inputTokens - p.outputTokens);
    reductions.set(key, group);
  }
  const requestKey = (event: EconomyEvent) => event.requestId === null ? null : JSON.stringify([event.projectId, event.sessionId, event.requestId, event.attemptId]);
  const observedRequests = new Set(usage.map(requestKey).filter(key => key !== null));
  const transformedRequests = new Set(transforms.map(requestKey).filter(key => key !== null && observedRequests.has(key)));
  const modeledCosts = usage.map(event => event.modelId === null ? null : event.payload.costUsd);
  return {
    schemaVersion: 1,
    period: { ...window, interval: '[since, until)' },
    sources: [...new Set(events.map(event => event.sourceId))],
    observedCalls: usage.filter(event => event.payload.observationScope !== 'session_delta').length,
    sessionUsageDeltas: usage.filter(event => event.payload.observationScope === 'session_delta').length,
    incompleteCalls: usage.filter(event => !event.payload.complete).length,
    unknownModels: usage.filter(event => event.modelId === null).length,
    coverage: { transformedObservedRequests: transformedRequests.size, observedRequestsWithId: observedRequests.size, scope: 'Imported or instrumented requests only; unobserved traffic is unknown.' },
    directInputReduction: [...reductions.values()].map(({ values, ...rest }) => ({ ...rest, tokens: sum(values) })),
    candidateTransformsNotCounted: events.filter(event => event.eventType === 'transform' && !event.payload.sent).length,
    modeledCostUsd: sum(modeledCosts),
    auxiliaryCostUsd: sum(usage.filter(event => event.payload.category !== 'primary').map(event => event.modelId === null ? null : event.payload.costUsd)),
    financialSavingsUsd: null,
    baseline: 'Unavailable: no comparable financial baseline was recorded.',
    taskMeasurement: 'Incomplete: instrumentation does not establish that every call of a task was captured.',
    groups: [...groups.entries()].map(([key, items]) => ({
      key, calls: items.length,
      input: sum(items.map(event => event.payload.inputTokens)),
      output: sum(items.map(event => event.payload.outputTokens)),
      cacheRead: sum(items.map(event => event.payload.cacheReadTokens)),
      cacheWrite: sum(items.map(event => event.payload.cacheWriteTokens)),
      costUsd: sum(items.map(event => event.modelId === null ? null : event.payload.costUsd)),
      evidence: [...new Set(items.map(event => event.evidence))],
      locations: [...new Set(items.map(event => event.executionLocation))],
    })),
  };
}

export type Report = ReturnType<typeof report>;
const display = (value: number | null): string => value === null ? 'unavailable' : String(value);
export function renderText(data: Report): string {
  return [
    'Julius — evidence-aware report',
    `Period: ${data.period.since} ≤ time < ${data.period.until} (${data.period.timezone})`,
    `Sources: ${data.sources.join(', ') || 'none'}`,
    `Observed calls: ${data.observedCalls}; incomplete: ${data.incompleteCalls}; unknown model: ${data.unknownModels}`,
    `Session usage deltas: ${data.sessionUsageDeltas} (not a known call count)`,
    `Coverage: ${data.coverage.transformedObservedRequests}/${data.coverage.observedRequestsWithId} observed requests with IDs transformed`,
    ...data.directInputReduction.map(item => `Direct input reduction: ${display(item.tokens.total)} tokens; known subtotal ${item.tokens.known}; ${item.scope}; ${item.evidence}; tokenizer ${item.tokenizer ?? 'unknown'}`),
    ...(data.directInputReduction.length ? [] : ['Direct input reduction: unavailable (no sent transformation evidence)']),
    `Modeled cost USD: ${display(data.modeledCostUsd.total)}; known subtotal: ${data.modeledCostUsd.known}`,
    `Observed auxiliary modeled cost USD: ${display(data.auxiliaryCostUsd.total)}`,
    `Estimated financial savings USD: unavailable. ${data.baseline}`,
    `Task measurement: ${data.taskMeasurement}`,
    ...data.groups.map(group => `${group.key}: calls=${group.calls}, input=${display(group.input.total)}, output=${display(group.output.total)}, evidence=${group.evidence.join(',')}`),
  ].join('\n');
}

function escapeHtml(value: string): string {
  return value.replace(/[&<>"']/g, char => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;', "'": '&#39;' }[char]!));
}
export function renderHtml(data: Report): string {
  return `<!doctype html><html lang="en"><meta charset="utf-8"><meta name="viewport" content="width=device-width"><meta http-equiv="Content-Security-Policy" content="default-src 'none'; style-src 'unsafe-inline'"><title>Julius evidence report</title><style>:root{color-scheme:light dark}body{font:16px system-ui;max-width:1000px;margin:3rem auto;padding:1rem}pre{white-space:pre-wrap;line-height:1.7}table{border-collapse:collapse;width:100%}td,th{padding:.7rem;text-align:left;border-bottom:1px solid #888}caption{text-align:left;font-weight:bold}</style><main><h1>Julius</h1><p>Local evidence report. No model calls. No external resources.</p><pre>${escapeHtml(renderText(data))}</pre><table><caption>Observed usage by model or selected dimension</caption><thead><tr><th scope="col">Group</th><th scope="col">Calls</th><th scope="col">Input</th><th scope="col">Output</th><th scope="col">Evidence</th></tr></thead><tbody>${data.groups.map(group => `<tr><th scope="row">${escapeHtml(group.key)}</th><td>${group.calls}</td><td>${display(group.input.total)}</td><td>${display(group.output.total)}</td><td>${escapeHtml(group.evidence.join(', '))}</td></tr>`).join('')}</tbody></table></main></html>`;
}

export function renderCsv(data: Report): string {
  const cell = (value: unknown) => `"${String(value ?? 'unavailable').replace(/^[=+@\-\t\r]/, "'$&").replaceAll('"', '""')}"`;
  return [['group', 'calls', 'input_tokens', 'output_tokens', 'modeled_cost_usd', 'evidence'], ...data.groups.map(group => [group.key, group.calls, group.input.total, group.output.total, group.costUsd.total, group.evidence.join(';')])].map(row => row.map(cell).join(',')).join('\n');
}

File: python/julius/reporting.py
"""Offline aggregates. Evidence remains separate from counterfactual claims."""

import csv
import html
import io
from collections import defaultdict
from collections.abc import Sequence
from typing import Any


def _sum(values: Sequence[int | float | None]) -> dict[str, Any]:
    known = sum(value for value in values if value is not None)
    unknown = sum(value is None for value in values)
    return {
        "known": known,
        "unknownRecords": unknown,
        "total": known if values and not unknown else None,
    }


def report(events: list[dict], window: dict, by: str = "model") -> dict[str, Any]:
    if by not in ("model", "client", "category"):
        raise ValueError("Group must be model, client, or category")
    usage = [event for event in events if event["eventType"] == "usage"]
    transforms = [
        event for event in events if event["eventType"] == "transform" and event["payload"]["sent"]
    ]
    groups: dict[str, list[dict]] = defaultdict(list)
    reductions: dict[tuple, list[int | None]] = defaultdict(list)
    for event in usage:
        key = (
            (event.get("modelId") or "unknown")
            if by == "model"
            else (event["clientId"] if by == "client" else event["payload"]["category"])
        )
        groups[key].append(event)
    for event in transforms:
        payload = event["payload"]
        key = (payload["scope"], event["evidence"], payload["tokenizer"], event["modelId"])
        before, after = payload["inputTokens"], payload["outputTokens"]
        reductions[key].append(None if before is None or after is None else before - after)

    def request_key(event: dict) -> tuple | None:
        return (
            (event["projectId"], event["sessionId"], event["requestId"], event["attemptId"])
            if event["requestId"] is not None
            else None
        )

    observed = {request_key(event) for event in usage} - {None}
    transformed = {request_key(event) for event in transforms} & observed

    def costs(items: list[dict]) -> list[float | None]:
        return [event["payload"]["costUsd"] if event["modelId"] else None for event in items]

    return {
        "schemaVersion": 1,
        "period": {**window, "interval": "[since, until)"},
        "sources": sorted({event["sourceId"] for event in events}),
        "observedCalls": sum(
            event["payload"].get("observationScope") != "session_delta" for event in usage
        ),
        "sessionUsageDeltas": sum(
            event["payload"].get("observationScope") == "session_delta" for event in usage
        ),
        "incompleteCalls": sum(not event["payload"]["complete"] for event in usage),
        "unknownModels": sum(event["modelId"] is None for event in usage),
        "coverage": {
            "transformedObservedRequests": len(transformed),
            "observedRequestsWithId": len(observed),
            "scope": "Imported or instrumented requests only; unobserved traffic is unknown.",
        },
        "directInputReduction": [
            {
                "scope": key[0],
                "evidence": key[1],
                "tokenizer": key[2],
                "modelId": key[3],
                "tokens": _sum(values),
            }
            for key, values in reductions.items()
        ],
        "candidateTransformsNotCounted": sum(
            event["eventType"] == "transform" and not event["payload"]["sent"] for event in events
        ),
        "modeledCostUsd": _sum(costs(usage)),
        "auxiliaryCostUsd": _sum(
            costs([e for e in usage if e["payload"]["category"] != "primary"])
        ),
        "financialSavingsUsd": None,
        "baseline": "Unavailable: no comparable financial baseline was recorded.",
        "taskMeasurement": "Incomplete: instrumentation does not establish that every call of a task was captured.",
        "groups": [
            {
                "key": key,
                "calls": len(items),
                **{
                    name: _sum([item["payload"][field] for item in items])
                    for name, field in (
                        ("input", "inputTokens"),
                        ("output", "outputTokens"),
                        ("cacheRead", "cacheReadTokens"),
                        ("cacheWrite", "cacheWriteTokens"),
                    )
                },
                "costUsd": _sum(costs(items)),
                "evidence": sorted({item["evidence"] for item in items}),
                "locations": sorted({item["executionLocation"] for item in items}),
            }
            for key, items in groups.items()
        ],
    }


def _display(value: Any) -> str:
    return "unavailable" if value is None else str(value)


def render_text(data: dict) -> str:
    period = data["period"]
    lines = [
        "Julius — evidence-aware report",
        f"Period: {period['since']} ≤ time < {period['until']} ({period['timezone']})",
        f"Sources: {', '.join(data['sources']) or 'none'}",
        f"Observed calls: {data['observedCalls']}; incomplete: {data['incompleteCalls']}; unknown model: {data['unknownModels']}",
        f"Session usage deltas: {data['sessionUsageDeltas']} (not a known call count)",
        f"Coverage: {data['coverage']['transformedObservedRequests']}/{data['coverage']['observedRequestsWithId']} observed requests with IDs transformed",
    ]
    lines.extend(
        f"Direct input reduction: {_display(item['tokens']['total'])} tokens; known subtotal {item['tokens']['known']}; {item['scope']}; {item['evidence']}; tokenizer {item['tokenizer'] or 'unknown'}"
        for item in data["directInputReduction"]
    )
    if not data["directInputReduction"]:
        lines.append("Direct input reduction: unavailable (no sent transformation evidence)")
    lines.extend(
        [
            f"Modeled cost USD: {_display(data['modeledCostUsd']['total'])}; known subtotal: {data['modeledCostUsd']['known']}",
            f"Observed auxiliary modeled cost USD: {_display(data['auxiliaryCostUsd']['total'])}",
            f"Estimated financial savings USD: unavailable. {data['baseline']}",
            f"Task measurement: {data['taskMeasurement']}",
        ]
    )
    lines.extend(
        f"{group['key']}: calls={group['calls']}, input={_display(group['input']['total'])}, output={_display(group['output']['total'])}, evidence={','.join(group['evidence'])}"
        for group in data["groups"]
    )
    return "\n".join(lines)


def render_csv(data: dict) -> str:
    output = io.StringIO(newline="")
    writer = csv.writer(output, quoting=csv.QUOTE_ALL)
    writer.writerow(
        ["group", "calls", "input_tokens", "output_tokens", "modeled_cost_usd", "evidence"]
    )
    for group in data["groups"]:
        row = [
            group["key"],
            group["calls"],
            group["input"]["total"],
            group["output"]["total"],
            group["costUsd"]["total"],
            ";".join(group["evidence"]),
        ]
        cells = [_display(value) for value in row]
        writer.writerow(
            [
                "'" + value if value.startswith(("=", "+", "-", "@", "\t", "\r")) else value
                for value in cells
            ]
        )
    return output.getvalue()


def render_html(data: dict) -> str:
    rows = "".join(
        f"<tr><th scope='row'>{html.escape(group['key'])}</th><td>{group['calls']}</td><td>{_display(group['input']['total'])}</td><td>{_display(group['output']['total'])}</td></tr>"
        for group in data["groups"]
    )
    return (
        "<!doctype html><html lang='en'><meta charset='utf-8'><meta name='viewport' content='width=device-width'>"
        "<meta http-equiv='Content-Security-Policy' content=\"default-src 'none'; style-src 'unsafe-inline'\">"
        "<title>Julius evidence report</title><style>:root{color-scheme:light dark}body{font:16px system-ui;max-width:1000px;margin:3rem auto;padding:1rem}pre{white-space:pre-wrap;line-height:1.7}table{border-collapse:collapse;width:100%}td,th{padding:.7rem;text-align:left;border-bottom:1px solid #888}</style>"
        f"<main><h1>Julius</h1><p>Local evidence report. No model calls or external resources.</p><pre>{html.escape(render_text(data))}</pre>"
        "<table><caption>Observed usage</caption><thead><tr><th scope='col'>Group</th><th scope='col'>Records</th><th scope='col'>Input</th><th scope='col'>Output</th></tr></thead>"
        f"<tbody>{rows}</tbody></table></main></html>"
    )
